app/controllers/locations_controller.rb

```diff
@@ -27,14 +27,40 @@ def index
       search_params["main_category"] = ""
     end
 
-    if !params["categories"] and !search_params["main_category"].empty?
-      if params["main_category_id"].present?
-        search_params["categories"] = [params["main_category_id"]]
-      end
+    @matched_category = match_keyword_to_category(search_params[:keyword])
+
+    if @matched_category
+      search_params[:categories] = [@matched_category.id]
+      search_params[:main_category] = @matched_category.parent&.name || @matched_category.name
+      @main_category_selected_name = search_params[:main_category]
+      @main_category_selected_id = @matched_category.parent&.id || @matched_category.id
+      search_params[:keyword] = nil
     else
-      search_params["categories"] = params["categories_ids"]
-    end
+      @main_category_selected_name = ""
+      @main_category_selected_id = nil
+      search_params[:categories] = []
+
+      if search_params[:keyword].present?
+        search_params[:main_category] = ""
+        params[:categories] = []
+      else
+        @main_category_selected_name = search_params[:main_category]
+        @main_category_selected_id = helpers.get_category_id_by_name(@main_category_selected_name) if @main_category_selected_name.present?
+
+        if !params["categories"] && !search_params["main_category"].empty?
```

Comment:
I see some times we use `params` and sometime `search_params` is there a reason for this?

Reply:
Yeah, I just followed the previous behavior in the code for how we handle the params here. A shallow copy of the params was made `search_params = params.dup` so we can modify the search_params without changing the original params.